File: 프로그래머스/lv1/12934. 정수 제곱근 판별/정수 제곱근 판별.js
function solution(n) {

    let sqrt = Math.sqrt(n);
    let result = Math.pow( (sqrt + 1) , 2);

    if( !Number.isInteger(sqrt) ) {
        return -1;
    }

    return result;

}
/*
    접근법 :
        1. n의 제곱근
        2. n이 정수인지 소수점인지 판별
        3. 정수면 (제곱근 + 1) 제곱
        4. 소수점이면 -1

    .sqrt()                     : 루트를 씌워 제곱근을 구해준다.
    .pow(대상값, 몇번제곱할래?) : 제곱이 필요할 때 사용하는 함수
    Number.isInteger()          : 매개변수로 들어온 값이 정수이니?


    다른 접근법 :
        정수는 1로 나누면 항상 나머지가 0이 된다.

    function solution(n) {

        let sqrt = Math.sqrt(n);
        let result = Math.pow( (sqrt + 1) , 2);

        if( sqrt % 1 === 0 ) {
            return result;
        }

        return -1;

    }
*/

File: 프로그래머스/lv1/12934. 정수 제곱근 판별/정수 제곱근 판별.java
class Solution {
    public long solution(long n) {

        double sqrt = Math.sqrt(n);
        long result = (long)Math.pow( (sqrt + 1), 2 );

        if(sqrt == Math.floor(sqrt)) {
            return result;
        }

        return -1;
    }
}
/*
    접근법 :
        - long타입으로 들어오고, long타입으로 반환 **
        - 제곱근, 제곱을 구하기 위해 Math class 사용
        - 자바에서 Math class 는 double 타입으로 반환

    .sqrt() : 소수점인지 아닌지 판별하려고 double로 받음
    .pow() : double 타입의 반환값을 메소드 반환값인 long타입으로 형변환


    다른 접근법 :
        좀더 간결한 코드는 없나..?
    
    - 간결하지만 알아보기 쉽다고는 볼 수 없을것 같다...
    class Solution {
        public long solution(long n) {
            
            double sqrt = Math.sqrt(n);

            return sqrt == Math.floor(sqrt) ? (long)Math.pow( (sqrt + 1), 2 ) : -1;

        }
    }
*/